### ISSUE TITLE
Build hangs when Concourse Docker container restarted

### ISSUE THREAD

**Issue by mbjelac** [concourse/concourse#1797]
This is an issue somewhat similar to another I'm having: #1796
Actually, I discovered this issue by trying to reconstruct that one.
They might have the same cause, or the other might be several issues in one. Anyway, here's the info, so you can check it out.

# Bug Report

* Concourse version: 3.5.0
* Deployment type (BOSH/Docker/binary): Docker 17.06.0-ce, build 02c1d87
* Infrastructure/IaaS: Linux (privately hosted virtual machine)
```
DISTRIB_ID=Ubuntu
DISTRIB_RELEASE=16.04
DISTRIB_CODENAME=xenial
DISTRIB_DESCRIPTION="Ubuntu 16.04.2 LTS"
NAME="Ubuntu"
VERSION="16.04.2 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.2 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```
* Browser (if applicable): Google Chrome 62.0.3202.89 (Official Build) (64-bit)
* Did this used to work? No, it happens consistently with older versions too.

## TL;DR

When I restart the Concourse containers (all of them, since I use `docker-compose`, but I guess the important one is the worker container), all jobs hang in the `preparing build` phase.

Fly reports the single worker as stalled. When I prune it, all goes back to normal.

## Reconstruction cookbook

- install Docker on a machine (see version above)
- start Docker service
- start Concourse containers with `docker-compose up -d` using docker-compose definition:
```yaml
version: '3.3'

services:
  concourse-db:
    image: postgres:9.5
    volumes: ["/srv/docker/containers/concourse-db/data:/database"]
    restart: unless-stopped
    logging:
      driver: json-file
      options:
        max-size: "10m"
        max-file: "5"
    environment:
      POSTGRES_DB: concourse
      POSTGRES_USER: concourse
      POSTGRES_PASSWORD: changeme
      PGDATA: /database

  concourse-web:
    image: concourse/concourse
    links: [concourse-db]
    command: web
    depends_on: [concourse-db]
    ports: ["7070:8080"]
    volumes: ["/etc/docker/container/concourse/keys/web:/concourse-keys"]
    restart: unless-stopped # required so that it retries until conocurse-db comes up
    logging:
      driver: json-file
      options:
        max-size: "10m"
        max-file: "10"
    environment:
      CONCOURSE_BASIC_AUTH_USERNAME: concourse
      CONCOURSE_BASIC_AUTH_PASSWORD: changeme
      CONCOURSE_EXTERNAL_URL: http://<ci-host>:7070
      CONCOURSE_POSTGRES_HOST: concourse-db
      CONCOURSE_POSTGRES_USER: concourse
      CONCOURSE_POSTGRES_PASSWORD: changeme
      CONCOURSE_POSTGRES_DATABASE: concourse

  concourse-worker:
    image: concourse/concourse
    privileged: true
    links: [concourse-web]
    depends_on: [concourse-web]
    command: worker
    volumes: 
      - "/etc/docker/container/concourse/keys/worker:/concourse-keys"
      - "/srv/docker/containers/concourse-worker/work-dir:/work-dir"
    restart: unless-stopped
    logging:
      driver: json-file
      options:
        max-size: "10m"
        max-file: "5"
    environment:
      CONCOURSE_TSA_HOST: concourse-web
      CONCOURSE_WORK_DIR: /work-dir
```
- create a pipeline
- start pipeline and verify they work as expected
- on the machine you installed Docker on, do *one* of (they all have the same results):
    - `docker-compose restart`
    - `docker-compose down` + `docker-compose up -d`
    - `docker-compose stop` + `docker-compose start`

## What happens?

- any build you start hangs:
```
---------------
preparing build
---------------
(checkmark) checking pipeline is not paused
(checkmark) checking job is not paused
(checkmark) discovering any new versions of ci-tools
(checkmark) discovering any new versions of source
(checkmark) waiting for a suitable set of input versions
(checkmark) checking max-in-flight is not reached
<waits forever>
```

## Check workers

- the single worker is stalled:
```
> fly -t ci workers
name          containers  platform  tags  team  state    version
b640b1adefed  0           linux     none  none  running  1.2


the following workers have not checked in recently:

name          containers  platform  tags  team  state    version
30d2bf6be64d  4           linux     none  none  stalled  1.2

these stalled workers can be cleaned up by running:

    fly -t ci prune-worker -w (name)
```

I don't know why Fly reports a new worker, there is only one worker registered. I guess its waiting for the stalled one to die so it can continue?

## Quickfix

- prune the stalled worker:
```
> fly -t ci prune-worker -w 30d2bf6be64d                           
pruned '30d2bf6be64d'                                            
                                                                 
> fly -t ci workers                                                
name          containers  platform  tags  team  state    version 
b640b1adefed  4           linux     none  none  running  1.2     
```

Builds start normally.

**Comment by francoisruty** [comment]
I'm encountering the exact same issue (using concourse 3.7.0 with docker-compose)

**Comment by vito** [comment]
Consolidating into https://github.com/concourse/concourse/issues/1457 - thanks for all the info!
